include/hipSYCL/sycl/info/param_traits.hpp

```diff
@@ -35,12 +35,12 @@ namespace hipsycl {
 namespace sycl {
 namespace info {
 
-template <typename T, T Param>
+template <typename Param>
 struct param_traits {};
 
 #define HIPSYCL_PARAM_TRAIT_RETURN_VALUE(T, param, ret_value) \
   template<> \
-  struct param_traits<T, param> \
+  struct param_traits<param> \
```

Comment:
I've just looked at the spec again to check what it says on `param_traits` - the only mention I find is this:

> Information query descriptors have been changed to structures under namespaces named accordingly. param_traits has been removed and the return type of an information query is now contained in the descriptor. [...]

from here: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:what-changed-between

However, I can't find any mention of how exactly the return type should be exposed in the struct. Arguably, that one quote might not even be normative since it just describes the changes as a changelog. 

Are you aware of any place where it says more?
I'm not sure it makes sense to do the required refactoring if the spec does not even clearly say anything. What do you think?

Reply:
Right... I didn't think this was standardised. Well, looking at what our friends in blue did:

https://github.com/intel/llvm/blob/f81499ecad464bd03d7f50e95004eeed1e85dbc9/sycl/include/sycl/info/info_desc.hpp#L26-L29

We should probably do something similar then... I'm thinking of removing `param_traits.hpp` and creating two macros in `info.hpp`, one for the plain case and the other for the templated case, just like their `__SYCL_PARAM_TRAITS_SPEC` and `__SYCL_PARAM_TRAITS_TEMPLATE_SPEC`. Help me choose some names:
`HIPSYCL_INFO_DESCRIPTOR(param, ret_value)` and `HIPSYCL_INFO_DESCRIPTOR_TEMPLATE(...)`?

Reply:
Well, I'd say that arguably it's not standardized.. That one sentence is only a changelog. So, we could adopt the point of view that this is not required. However, it is possible that, if this is discussed inside the SYCL WG, it is found that this is just an oversight, and we have to implement the type inside the descriptor classes.

If you want to make the transition, I agree with the approach from DPC++.

Do we actually need two macros?

```c++
#define HIPSYCL_DECLARE_INFO_DESCRIPTOR(param, ret_type) \
 struct param { using return_type = ret_type; };

template<int Dim>
HIPSYCL_DECLARE_INFO_DESCRIPTOR(max_group_sizes, range<Dim>)
```
?

Reply:
True. Solid.

I'd just rather argue this a definition, and so `HIPSYCL_DEFINE_INFO_DESCRIPTOR()`.

Reply:
> I've just looked at the spec again to check what it says on `param_traits` - the only mention I find is this:
> 
> > Information query descriptors have been changed to structures under namespaces named accordingly. param_traits has been removed and the return type of an information query is now contained in the descriptor. [...]
> 
> from here: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:what-changed-between
> 
> However, I can't find any mention of how exactly the return type should be exposed in the struct. Arguably, that one quote might not even be normative since it just describes the changes as a changelog.
> 
> Are you aware of any place where it says more? I'm not sure it makes sense to do the required refactoring if the spec does not even clearly say anything. What do you think?

I found this for `get_info()` from https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.members.device:

"Queries this SYCL device for information requested by the template parameter Param. The type alias Param::return_type must be defined in accordance with the info parameters in [Table 25](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.device.info) to facilitate returning the type associated with the Param parameter."

Reply:
In hindsight, I'm supper happy we decided to do the refactoring.
Same functionality (or just slightly more) in 58% of the lines of code. 🎉

Reply:
Yep, looks much better :) Thanks for all this work :)